File: car-rental-backend/app/schemas/rental.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class PickupRequest(BaseModel):
    mileage_start: int = Field(ge=0)
    fuel_level_start: Decimal = Field(ge=0, le=100)
    photo_urls: list[str] = Field(default_factory=list)
    client_signature_url: str | None = None


class ReturnRequest(BaseModel):
    mileage_end: int = Field(ge=0)
    fuel_level_end: Decimal = Field(ge=0, le=100)
    damage_notes: str | None = None
    damage_photo_urls: list[str] = Field(default_factory=list)
    extra_charges: Decimal = Field(default=Decimal("0"), ge=0)


class RentalPriceBreakdownResponse(BaseModel):
    base_price: Decimal
    fuel_surcharge: Decimal
    risk_multiplier: Decimal
    final_price: Decimal
    calculated_at: datetime

    model_config = {"from_attributes": True}


class RentalResponse(BaseModel):
    id: uuid.UUID
    reservation_id: uuid.UUID
    pickup_date: datetime
    return_date: datetime | None
    mileage_start: int
    mileage_end: int | None
    fuel_level_start: Decimal
    fuel_level_end: Decimal | None
    damage_notes: str | None
    employee_id: uuid.UUID
    price_breakdown: RentalPriceBreakdownResponse | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: car-rental-frontend/src/types/rental.ts
export interface UserRentalVehicle {
  id: string;
  brand: string;
  model: string;
  year: number;
  license_plate: string;
  image_url: string | null;
}

export interface UserRentalItem {
  id: string;
  reservation_id: string;
  vehicle: UserRentalVehicle;
  pickup_date: string;
  return_date: string | null;
  status: string;
  total_price: string;
  final_price: string | null;
  created_at: string;
}

export interface PaginatedUserRentalsApi {
  items: UserRentalItem[];
  total: number;
  offset: number;
  limit: number;
}

export interface RentalPriceBreakdown {
  base_price: string;
  fuel_surcharge: string;
  risk_multiplier: string;
  final_price: string;
  calculated_at: string;
}

export interface RentalResponse {
  id: string;
  reservation_id: string;
  pickup_date: string;
  return_date: string | null;
  mileage_start: number;
  mileage_end: number | null;
  fuel_level_start: string;
  fuel_level_end: string | null;
  damage_notes: string | null;
  employee_id: string;
  price_breakdown: RentalPriceBreakdown | null;
  created_at: string;
}

export interface AdminReservationUser {
  id: string;
  email: string;
  first_name: string;
  last_name: string;
}

export interface AdminReservationVehicle {
  id: string;
  brand: string;
  model: string;
  license_plate: string;
}

export interface AdminReservationItem {
  id: string;
  user: AdminReservationUser;
  vehicle: AdminReservationVehicle;
  start_date: string;
  end_date: string;
  status: string;
  total_price: string;
  created_at: string;
}

export interface PaginatedAdminReservationsApi {
  items: AdminReservationItem[];
  total: number;
  offset: number;
  limit: number;
}
